### DISCUSSION TITLE
LDM Notes for May 13th, 2024

### DISCUSSION THREAD

**Discussion by 333fred** [dotnet/csharplang#8118] [LDM Notes]
https://github.com/dotnet/csharplang/blob/main/meetings/2024/LDM-2024-05-13.md

## Agenda

- First-class span types questions
    - Delegate conversions
    - Variant conversion existence
- Overload resolution priority questions
    - Attribute shape and inheritance
    - Extension overload resolution

**Comment by hamarb123** [comment]
> We could generate a `<PrivateImplementationDetails>` method for converting the span. On platforms with built-in `ReadOnlySpan<T>` support, and in particular .NET Core 2.2 and later, this seems like it would work fine. However, we're concerned about other platforms; users can bring their own `ReadOnlySpan<T>` implementations, and there's no guarantee that whatever bit-blit strategy the compiler choses would be safe for that implementation. We also aren't prepared to make strong statements on whether this would always be safe on the .NET Framework implementation of `ReadOnlySpan<T>`.

Just a note (for people who aren't aware) that bit-blit is not required for .NET Core 2.1+ and .NET Standard 2.1 (specifically, whenever `MemoryMarshal.CreateReadOnlySpan` is available).
The following could be used, which is the same implementation as in the runtime effectively (but using `MemoryMarshal` instead of the internal APIs, since that's how they're exposed publicly):
```csharp
class <PrivateImplementationDetails>
{
    public static ReadOnlySpan<T> CastUp<TDerived>(ReadOnlySpan<TDerived> items) where TDerived : class?, T
    {
        return MemoryMarshal.CreateReadOnlySpan(ref Unsafe.As<TDerived, T>(ref MemoryMarshal.GetReference(items)), items.Length);
    }
}
```

Runtime implementation:
https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/ReadOnlySpan.cs#L226-L229

**Reply by CyrusNajmabadi** [reply]
@stephentoub ?

**Reply by 333fred** [reply]
Yes, that's correct, it's what Stephen said during the meeting, and is in fact what I meant when I said it would be fine on .NET Core 2.2 and higher. It's the other platforms where we are concerned 🙂.
